compiler/rustc_hir_analysis/src/variance/mod.rs

```diff
@@ -112,10 +112,8 @@ fn variance_of_opaque(tcx: TyCtxt<'_>, item_def_id: LocalDefId) -> &[ty::Varianc
         fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
             match t.kind() {
                 ty::Alias(_, ty::AliasTy { def_id, substs, .. })
-                    if matches!(
-                        self.tcx.def_kind(*def_id),
-                        DefKind::OpaqueTy | DefKind::ImplTraitPlaceholder
-                    ) =>
+                    if matches!(self.tcx.def_kind(*def_id), DefKind::OpaqueTy)
+                        || self.tcx.is_impl_trait_in_trait(*def_id) =>
```

Comment:
Can you split this into two branches, `ty::Alias(ty::Opaque)` and `ty::Alias(..) if is_impl_trait_in_trait`?

Also, I'm not actually sure if we need to do this for synthetic associated items...? Could use a `FIXME(-Z...)` that asks whether this is necessary at all for RPITITs.